### DISCUSSION TITLE
turn off alignment completely ?

### DISCUSSION THREAD

**Discussion by aaamoeder** [jomjol/AI-on-the-edge-device#1338] [Q&A] [answered]
I was wondering if it's possible to completely skip the alignment process ?
My esp is fixed firmly and images are positioned 100% exactly the same each time so there is no real need for alignment ?

Actually the image has very little points for alignment resulting in the alignment often tilting the image while it was perfectly aligned to begin with.

thanks in advance !

**Comment by jomjol** [comment]
No this not possible with this firmware. The flow is fixed inside the firmware

**Comment by caco3** [comment]
> Actually the image has very little points for alignment resulting in the alignment often tilting the image while it was perfectly aligned to begin with.

Can you add a screenshot.
Usually it helps to put on stickers creating some good contrast

**Reply by aaamoeder** [reply]
> > Actually the image has very little points for alignment resulting in the alignment often tilting the image while it was perfectly aligned to begin with.
> 
> Can you add a screenshot. Usually it helps to put on stickers creating some good contrast

![Screenshot_20221113-012912](https://user-images.githubusercontent.com/20947118/201499941-fee7e6ed-017c-4e13-80aa-1c6846bfe59a.png)

**Reply by jomjol** [reply]
You can put two little white sticker with a black dor or something similar on top of the glass (e.g. between the two number, one left, one right).
This is much easier, than reprogramming the firmware.

**Reply by caco3** [reply]
And I guess this screenshot is before the alignment, right?
Also, the alignment marks (blue) should not overlap the numbers (red)!

**Reply by aaamoeder** [reply]
> You can put two little white sticker with a black dor or something similar on top of the glass (e.g. between the two number, one left, one right). This is much easier, than reprogramming the firmware.

This wont work since there is a plexiglass casing around the meter through which I'm taking the pictures (case is sealed so I can't open it)
Putting stickers on the plexiglass messes with the focussing big time because it's too close to the camera.

> And I guess this screenshot is before the alignment, right? Also, the alignment marks (blue) should not overlap the numbers (red)!

This was actually after..
I've now repositioned the esp so there is a small logo visible and it appears to help for the time being.

**Comment by aaamoeder** [comment]
I've tried to fix it but still, lots of times the alignment algo screws up.. See added images..

**PRE alignment**
![prealignment](https://user-images.githubusercontent.com/20947118/203819643-78f12446-d648-4e9a-95af-8ce4ba36b116.png)

**POST alignment**
![postalignment](https://user-images.githubusercontent.com/20947118/203819656-7f4b71a7-3604-4f5f-b11c-8621e5cc4189.png)

**alignment mark 1**
![alignmentmark1](https://user-images.githubusercontent.com/20947118/203819673-89fe9ab3-0eec-4326-a01f-08518092f758.png)

**alignment mark 2**
![alignmentmark2](https://user-images.githubusercontent.com/20947118/203819678-5660be12-3b4c-4e01-990b-99a1dc5a3a16.png)

**Reply by jomjol** [reply]
Why don't you put a little white sticker with some black dot between the numbers? There should be enough space

**Reply by aaamoeder** [reply]
the camera is on a closed acrilic case which is sealed by the utility supplier 
Putting a sticker on the acrilic doesn't work since it's only ~ 1cm from the camera while the numbers are ~6cm from the camera so that causes all kinds of focus issues.

**Comment by jomjol** [comment]
Okay got the point. Then currently there is not a good solution for your specific problem. Sorry for this!

**Reply by aaamoeder** [reply]
Would selecting more/different letters help with alignment or is there no real difference?

**Reply by caco3** [reply]
You should put the alignment marks as far apart as possible to increase alignment accuracy.

**Comment by nareddyt** [comment]
+1 to completely disable alignment. It seems kinda redundant given that the ESP camera is in a fixed position. In addition to taking extra processing time, it also messes up randomly and tilts the image (similar to screenshots above posted by @aaamoeder ).

I was able to get better results by:

1. Changing the alignment references to use dots instead of text. i.e. I added a reference on the blue dots at the bottom of the image, and it works much better.
2. In the configuration (expert mode), I changed `Alignment.AlignmentAlgo=Fast`. IIUC based on the help text below, it prevents drastic changes to the alignment after the initial startup run unlike the default value? @jomjol can you confirm?

> "Default" = use only R-Channel, "HighAccuracy" = use all Channels (RGB, 3x slower), "Fast" (First time RGB, then only check if image is shifted)

![image](https://user-images.githubusercontent.com/11142171/203855130-e625ecc1-241b-44fd-a434-03bf2162f061.png)

**Reply by caco3** [reply]
As I understand it, it takes the photo, then searches for the reference patterns and aligns the photo so the marks on the photo match the coordinates of the reference images.

It might be that when you used text as mark, that it found other text patterns which looked similar enough thus caused a false positive. Using unique references should avoid this (like your blue dots).

**Comment by aaamoeder** [comment]
Would selecting some number from the top meter (which is non functional so always the same) work ? or would it recognize the same number below and get confused ?
Not quite sure how the alignment algo actually works ....

**Reply by caco3** [reply]
I would think that any pattern which **does not change** and is **unique** should work.
So i would not use the top meter as it has the same pattern as the values below.

**Comment by caco3** [comment, marked as answer]
This just got implemented in https://github.com/jomjol/AI-on-the-edge-device/pull/1867

**Reply by nliaudat** [reply]
in PR https://github.com/jomjol/AI-on-the-edge-device/pull/1924 in fact
